### PR TITLE
feat: add wasi preview2 shim for browser and nodejs

### PR DESCRIPTION
**Pull request by eduardomourar**
This a basic shim for WASI (preview2) in JavaScript. The browser part is almost as is from https://github.com/guybedford/wasi-command-export-printf/tree/main/wasi and for Node.js it implements a small percentage of the API. Keep in mind that this is packaged as an ES module to be consumed for both environments without any workaround or hack.

### CODE REVIEW

File: packages/wasi-common/package.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+    "name": "@bytecodealliance/wasi-common",
+    "version": "0.0.1",
+    "description": "WASI implementation in JavaScript for the Browser and Node.js",
+    "author": "Guy Bedford, Eduardo Rodrigues<16357187+eduardomourar@users.noreply.github.com>",
+    "type": "module",
+    "main": "./index.js",
+    "browser": "./browser/index.js",
```
Comment (guybedford) [inline packages/wasi-common/package.json:8]:
```suggestion
    "exports": {
      "node": "./nodejs/index.js",
      "default": "./browser/index.js"
    }
```

File: packages/wasi-common/package.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{
+    "name": "@bytecodealliance/wasi-common",
```
Comment (guybedford) [inline packages/wasi-common/package.json:2]: Sounds good, but perhaps let's make it release-specific and note that this is not a "proper" implementation but a "shim" -

```suggestion
    "name": "@bytecodealliance/preview2-shim",
```

**Review by guybedford — APPROVED** [review]
Thanks for putting this together, it's almost entirely identical to what I've been using myself here.

Definitely agreed on the overall direction, just see the few comments.

Happy to merge and iterate from there.

File: packages/wasi-common/nodejs/exit.js [hunk 3]
```diff
@@ -0,0 +1,4 @@
+export function exit(status) {
+  console.log(`[exit] Exit: ${JSON.stringify(status)}`);
+  process.exit(1);
```
Comment (guybedford) [inline packages/wasi-common/nodejs/exit.js:3]: This should probably be code-specific:

```suggestion
  process.exit(status.tag === 'err' ? 1 : 0);
```

File: packages/wasi-common/types/wasi-exit.d.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export namespace WasiExit {
+  export function exit(status: Result<void, void>): void;
```
Comment (guybedford) [inline packages/wasi-common/types/wasi-exit.d.ts:2]:
```suggestion
  export function exit(status: { tag: 'err' | 'ok' }): void;
```

**Comment by guybedford** [comment]
We should probably be generating all the types directly from the WASI wit definitions as well to ensure we're in sync.

**Comment by eduardomourar** [comment]
I have made a commit with the experimental outbound HTTP request work I have been doing (related to this other PR [here](https://github.com/bytecodealliance/preview2-prototyping/pull/85)). I guess I should revert that, right?

**Comment by guybedford** [comment]
I'm more than happy to just ship things in any state here, as it is highly experimental and also because it is just a shim - so any iterative improvement in terms of things working will always be a win.

I'm tempted to post a first npm publish later this afternoon, unless you can think of any reasons not to!

**Comment by eduardomourar** [comment]
> I'm more than happy to just ship things in any state here, as it is highly experimental and also because it is just a shim - so any iterative improvement in terms of things working will always be a win.
> 
> 
> 
> I'm tempted to post a first npm publish later this afternoon, unless you can think of any reasons not to!

Publishing to NPM will actually make my life so much easier!

**Pull request merged**